### PR TITLE
Compile with -warnings-as-errors in CI (incl. StrictSendability)

### PR DESCRIPTION
**Pull request by simonjbeaumont**
### Motivation

We were not compiling with `-warnings-as-errors` in CI because some of the reference source code contained deprecation annotations. This meant that we are not getting CI failures for other warnings that we'd like to be clean of; specifically strict concurrency. However, in addition to the file-based reference tests, we now have in-memory snippet-based reference tests, which contain tests for deprecated OpenAPI schemas and operations.

### Modifications

- Remove deprecated tests from file-based reference test.
- Clean up warnings under StrictConcurrency.
- Compile with -warnings-as-errors in CI.
- Opt into StrictConcurrency and `-warnings-as-errors` locally with environment variable.

### Result

- CI will catch sendability warning regressions.
- Developing locally with `SWIFT_OPENAPI_STRICT_CONCURRENCY` will enable the same behaviour.

### Test Plan

- Added a canary commit to the end of this PR with a sendability issue, which will be reverted before merging.

### CODE REVIEW

**Comment by simonjbeaumont** [comment]
Canary commit shows that the CI will help us catch the warnings:

```
[737/761] Compiling OpenAPIGeneratorReferenceTests CompatabilityTest.swift
/code/Tests/OpenAPIGeneratorReferenceTests/CompatabilityTest.swift:213:42: error: capture of 'diagnosticsCollector' with non-sendable type 'RecordingDiagnosticCollector' in a `@Sendable` closure
                            diagnostics: diagnosticsCollector
                                         ^
/code/Tests/OpenAPIGeneratorReferenceTests/CompatabilityTest.swift:335:21: note: class 'RecordingDiagnosticCollector' does not conform to the 'Sendable' protocol
private final class RecordingDiagnosticCollector: DiagnosticCollector {
                    ^
...
error: fatalError
1

I'll now revert the canary commit for review.

File: Package.swift [hunk 1]
```diff
@@ -23,6 +24,15 @@ swiftSettings.append(
     // Require `any` for existential types.
     .enableUpcomingFeature("ExistentialAny")
 )
+
+// Strict concurrency is enabled in CI; use this environment variable to enable it locally.
+if ProcessInfo.processInfo.environment["SWIFT_OPENAPI_STRICT_CONCURRENCY"].flatMap(Bool.init) ?? false {
+    #warning("Compiling with Strict Concurrency")
+    swiftSettings.append(contentsOf: [
+        .enableExperimentalFeature("StrictConcurrency"),
```
Comment (czechboy0) [inline Package.swift:32]: Do we need this one if we already have this one? https://github.com/apple/swift-openapi-generator/pull/281/files#diff-cae86870418d35877d5f11c0ea0b536fdec37236d6ea5c0951eec4723f508ccfR15

Comment (simonjbeaumont) [reply Package.swift:32]: This isn't for CI, this is for local development, e.g. in Xcode.

Comment (czechboy0) [reply Package.swift:32]: Gotcha, that's good then.

**Review by czechboy0 — APPROVED** [review]
:shipit:

**Pull request merged**